Sources/NIOHTTP2/HTTP2CommonInboundStreamMultiplexer.swift

```diff
@@ -321,3 +406,116 @@ extension HTTP2CommonInboundStreamMultiplexer {
         }
     }
 }
+
+extension HTTP2CommonInboundStreamMultiplexer {
+    func setChannelContinuation(_ streamChannels: any ChannelContinuation) {
+        self.channel.eventLoop.assertInEventLoop()
+        self.streamChannelContinuation = streamChannels
+    }
+}
+
+/// `ChannelContinuation` is used to generic async-sequence-like objects to deal with `Channel`s. This is so that they may be held
+/// by the `HTTP2ChannelHandler` without causing it to become generic itself.
+internal protocol ChannelContinuation {
+    func yield(channel: Channel)
+    func finish()
+    func finish(throwing error: Error)
+}
+
+
+/// `StreamChannelContinuation` is a wrapper for a generic `AsyncThrowingStream` which holds the inbound HTTP2 stream channels.
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+struct StreamChannelContinuation<Output>: ChannelContinuation {
+    private var continuation: AsyncThrowingStream<Output, Error>.Continuation
+    private let inboundStreamInititializer: (Channel) -> EventLoopFuture<Output>
+
+    private init(
+        continuation: AsyncThrowingStream<Output, Error>.Continuation,
+        inboundStreamInititializer: @escaping (Channel) -> EventLoopFuture<Output>
+    ) {
+        self.continuation = continuation
+        self.inboundStreamInititializer = inboundStreamInititializer
+    }
+
+    /// `initialize` creates a new `StreamChannelContinuation` object and returns it along with its backing `AsyncThrowingStream`.
+    /// The `StreamChannelContinuation` provides access to the inbound HTTP2 stream channels.
+    ///
+    /// - Parameters:
+    ///   - inboundStreamInititializer: A closure which initializes the newly-created inbound stream channel and returns a generic.
+    ///   The returned type corresponds to the output of the channel once the operations in the initializer have been performed.
+    ///   For example an `inboundStreamInititializer` which inserts handlers before wrapping the channel in a `NIOAsyncChannel` would
+    ///   have a `Output` corresponding to that `NIOAsyncChannel` type. Another example is in cases where there is
+    ///   per-stream protocol negotiation where `Output` would be some form of `NIOProtocolNegotiationResult`.
+    static func initialize(
+        with inboundStreamInititializer: @escaping (Channel) -> EventLoopFuture<Output>
+    ) -> (StreamChannelContinuation<Output>, NIOHTTP2InboundStreamChannels<Output>) {
+        var continuation: AsyncThrowingStream<Output, Error>.Continuation? = nil
+        let stream = AsyncThrowingStream { continuation = $0 }
```

Comment:
NIT: In Swift 5.9+ we are going to have a method on `AsyncStream` which makes this easier called `makeStream`. This method returns the stream and the continuation. Instead of doing this here could we add this extension here ourselves surrounded with `#if` that guard the Swift version so that we can remove it in the future.

The extensions looks like this
```
extension AsyncThrowingStream {
  public static func makeStream(
      of elementType: Element.Type = Element.self,
      throwing failureType: Failure.Type = Failure.self,
      bufferingPolicy limit: Continuation.BufferingPolicy = .unbounded
  ) -> (stream: AsyncThrowingStream<Element, Failure>, continuation: AsyncThrowingStream<Element, Failure>.Continuation) where Failure == Error {
    var continuation: AsyncThrowingStream<Element, Failure>.Continuation!
    let stream = AsyncThrowingStream<Element, Failure>(bufferingPolicy: limit) { continuation = $0 }
    return (stream: stream, continuation: continuation!)
  }
}
```